bin/plugin/commands/benchmark.js

```diff
@@ -0,0 +1,231 @@
+ * @property {number} requests    A number of requests to perform.
+ * @property {string} file        A path to a file with URLs to benchmark.
+ * @property {string} output      An output format.
+ */
+exports.options = [
+	{
+		argname: '-u, --url <url>',
+		description: 'An URL to run benchmark tests for',
+	},
+	{
+		argname: '-c, --concurrency <concurrency>',
+		description: 'Number of multiple requests to make at a time',
+		defaults: 1,
+	},
+	{
+		argname: '-n, --requests <requests>',
+		description: 'Number of requests to perform',
+		defaults: 1,
+	},
+	{
+		argname: '-f, --file <file>',
+		description: 'File with URLs to run benchmark tests for',
+	},
+	{
+		argname: '-o, --output <output>',
+		description: 'Output format: csv or console',
+		defaults: 'console',
+	},
+];
+
+/**
+ * Runs http benchmarks for an URL.
+ *
+ * @param {BenchmarkCommandOptions} opt Command options.
+ */
+exports.handler = async ( opt ) => {
+	const { concurrency: connections, requests: amount } = opt;
+	const results = [];
+
+	for await ( const url of getURLs( opt ) ) {
+		const { responseTimes, metrics } = await benchmarkURL( {
+			url,
+			connections,
+			amount,
+		} );
+
+		results.push( [ url, responseTimes, metrics ] );
+	}
+
+	outputResults( opt, results );
+};
+
+/**
+ * Generates URLs to benchmark based on command arguments. If both "<url>" and "<file>" arguments
+ * are passed to the command, then both will be used to generate URLs.
+ *
+ * @param {BenchmarkCommandOptions} opt Command options.
+ */
+async function* getURLs( opt ) {
+	if ( !! opt.url ) {
+		yield opt.url;
+	}
+
+	if ( !! opt.file ) {
+		const rl = readline.createInterface( {
+			input: fs.createReadStream( opt.file ),
+			crlfDelay: Infinity,
+		} );
+
+		for await ( const url of rl ) {
+			if ( url.length > 0 ) {
+				yield url;
+			}
+		}
+	}
+}
+
+/**
+ * Benchmarks an URL and returns response time and server-timing metrics for every request.
+ *
+ * @param {BenchmarkOptions} params Benchmark parameters.
+ * @return {BenchmarkResults} Response times and metrics arrays.
+ */
+function benchmarkURL( params ) {
+	const metrics = {};
+	const responseTimes = [];
+
+	const onHeaders = ( { headers } ) => {
+		const responseMetrics = getServerTimingMetricsFromHeaders( headers );
+		Object.entries( responseMetrics ).forEach( ( [ key, value ] ) => {
+			metrics[ key ] = metrics[ key ] || [];
+			metrics[ key ].push( +value );
+		} );
+	};
+
+	const onResponse = ( statusCode, resBytes, responseTime ) => {
+		responseTimes.push( responseTime );
+	};
+
+	const instance = autocannon( {
+		...params,
+		setupClient( client ) {
+			client.on( 'headers', onHeaders );
+			client.on( 'response', onResponse );
+		},
+	} );
+
+	const onStop = instance.stop.bind( instance );
+	process.once( 'SIGINT', onStop );
+
+	return new Promise( ( resolve ) => {
+		instance.on( 'done', () => {
+			process.off( 'SIGINT', onStop );
+			resolve( { responseTimes, metrics } );
+		} );
+	} );
+}
```

Comment:
There's currently no error handling in this function. When calling this e.g. with a URL that doesn't exist (or basically any more realistic error that could happen), the table still gets a column for that URL contain just an `NaN` average time. It would be better to catch request failures or 400/500 responses and maybe log warnings for those.

I would say results for a URL where any errors occurred should not be included in the table (because if e.g. 2 out of 9 requests fail, the data would be inconsistent compared to the other URLs), and the warning should indicate that the user should run the check for those again or investigate what went wrong.

Reply:
> ... It would be better to catch request failures or 400/500 responses and maybe log warnings for those ...
> ... results for a URL where any errors occurred should not be included in the table ...

I don't think that the status code really matters for us. This tool is intended to benchmark the response time, not the correct behaviour of the site. If the site returns an error code, it still spends some time to process the request, right? Anyway we are going to use it for controlled set of pages where errors are not expected 🤓 

Added the completion rate to the results table, just in case something goes wrong suddenly.

Reply:
That sounds good, thanks!